### ISSUE TITLE
Detect that pgbouncer is running on current connection?

### ISSUE THREAD

**Issue by SamSaffron** [pgbouncer/pgbouncer#249]
Further to: 

https://github.com/rails/rails/issues/31190#issuecomment-346243803 

One thing that would help immensely with pgbouncer was if there was a way to detect easily if pgbouncer is running from the client. 

I was wondering if you would be open to adding "pgbouncer" to the version string. So this worked?

```
select version()
 PostgreSQL 9.5.5 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.2) 5.4.0 20160609, 64-bit, pgbouncer
```

or perhaps some other magic to consistently detect out of the box if a client is using pgbouncer or not?

**Comment by SamSaffron** [comment]
This could be achieved if we were guaranteed to have something like:

```
set session "pgbouncer.session" = '1';
```

run every time. (without needing special config and an opt-out option)

**Comment by matthewd** [comment]
Bonus points if we can also identify the pooling mode 🙏

**Comment by petere** [comment]
You could connect to the `pgbouncer` database and execute a command there and see if you get a familiar result back.

**Comment by SamSaffron** [comment]
@petere the trouble though is that the framework can not handle that transparently. 

As an app developer I can simply "disable feature X" or "disable feature Y", but the problem is that I need to remember what to disable and communicating this to every framework consumer is very hard. 

If we establish a standard here that pgbouncer follows then we can easily amend the framework to "feature detect" that pgbouncer is in place and in transaction mode without connecting to the pgbouncer database which we have no permission to.

**Comment by matthewd** [comment]
> without connecting to the pgbouncer database which we have no permission to

Yeah, the permission problem seems significant there. It's also pretty high-traffic... as a framework author, I'm okay with spending one RTT during connection establishment / configuration to identify pgbouncer... but something on the scale of a full separate connection seems too heavy a cost to inflict on everyone.

**Comment by petere** [comment]
I think a problem would be that pgbouncer is supposed to be a transparent connection proxy, so if you inject something that makes it nontransparent, then there will be other users who are going to be upset.

What is the reason you want to detect pgbouncer anyway?

And how should it behave if there are multiple pgbouncer hops?

**Comment by eulerto** [comment]
Also, I would suggest @petere solution [1]. The problem is that the version is printed as NOTICE instead of row. I created a PR #254 to allow creating a view for SHOW VERSION.

[1] http://peter.eisentraut.org/blog/2015/03/25/retrieving-pgbouncer-statistics-via-dblink/

**Comment by matthewd** [comment]
@petere pgbouncer applies a bunch of caveats on what you can safely do on the connection (depending on its pooling mode); we want to detect it so that we can avoid those features.

For multiple hops, my *ideal* behaviour would be that pgbouncer use the same protocol to identify that it's talking to another pgbouncer upstream, and then report the lowest-common-denominator pooling mode. But even without that, being able to automatically detect simple configurations would be a huge improvement, even if people with multi-layer setups still needed to provide manual configuration.

It's already hugely non-transparent if you happen to trip over a feature not supported by the active pooling mode, so comparatively, having it respond differently -- to something that a real Postgres server would never support -- doesn't seem horribly jarring.

@SamSaffron's idea of an automatic session variable feels *pretty* safe, just through the namespacing... but a more extreme alternative would be to detect a query text of exactly "PGBOUNCER", say (i.e., definitely not currently, nor ever going to be, a real SQL keyword). It could even still pass the query along, and just add an extra note to the error on its way back out.

**Comment by PJMODOS** [comment]
> It's already hugely non-transparent if you happen to trip over a feature not supported by the active pooling mode, so comparatively, having it respond differently -- to something that a real Postgres server would never support -- doesn't seem horribly jarring.
>
> @SamSaffron's idea of an automatic session variable feels pretty safe, just through the namespacing... but a more extreme alternative would be to detect a query text of exactly "PGBOUNCER", say (i.e., definitely not currently, nor ever going to be, a real SQL keyword). It could even still pass the query along, and just add an extra note to the error on its way back out.

We definitely wouldn't want to add query parsing since that would mean large architectural change of pgbouncer. Currently pgbouncer does not know anything about the commands passed through and often does not even deal with whole packet being sent. Changing this would have profound impact on both architecture and performance of pgbouncer in a way we don't want. It's same reason as why we won't do anything like read only query detection for example.

Now the session variable is relatively reasonably way of enabling users to do this. It could be done in similar fashion `reset_query`. We could add some `on_connect_query` parameter which would be empty by default and if it's set to something it would be sent after connection is made to server (you can then recommend users setting it to something like `SET pgbouncer.mode = 'transaction'` for your framework).

**Comment by eulerto** [comment]
> Now the session variable is relatively reasonably way of enabling users to do this. It could be done in similar fashion reset_query. We could add some on_connect_query parameter which would be empty by default and if it's set to something it would be sent after connection is made to server (you can then recommend users setting it to something like SET pgbouncer.mode = 'transaction' for your framework).

It doesn't seem flexible. If you have two pools (one `transaction` and another `session`), you can't decide the right value for `on_connect_query` (unless that parameter is per pool).

Another approach is to inject those custom options (such as `pgbouncer.version` and `pgbouncer.pool_mode`) after the connection is established if a configuration parameter (say `expose_pgbouncer`) is set (default is false). This parameter could also be per pool. Someone might want to be even more flexible by choosing those custom parameters. In this case, `expose_pgbouncer` would have a list as value.

**Comment by matthewd** [comment]
My use of 'exactly' wasn't accidental -- we're talking about a memcmp, not a full parser -- but I understand still not wanting to add anything to the "main" pipeline. We could move to a more obscure alternative like describe, but I assume your objection would remain.

My hope is for a behaviour that can be on by default: if we're going to require users to change a config setting, we might as well tell do it on our side. 

Could we supply a connect-time option of say `pgbouncer.mode='detect'`, and if you see that exact name+value, you change it to reflect the truth? 

Our only current mechanism for automatically determining which PostgreSQL features are available is the server version number -- but by randomly breaking features that are documented [by PG] to work, PgBouncer makes that very difficult. Is there *any* recommended way for a framework to selectively avoid the features that it breaks, while still being able to use them on a direct Postgres connection?

**Comment by mikebell90** [comment]
One example of a use case - and be gentle here, my ignorance is high is - https://github.com/flyway/flyway/issues/2895

* We provide a library that preps and configs our Datasource
* We call Flyway
* Assuming that fellow has a point (talking to dbas shortly about this), it would "nice" to be able to "sniff" if PGBOUNCER were being used, since I'm pretty sure the Flyway guys will reasonably say "that's an app config choice" if they change their code.

**Comment by eulerto** [comment]
@mikebell90 create a pool for migration whose `pool_mode = session`.

`flyway = host=192.168.1.2 port=5432 dbname=foobar pool_mode=session`

**Comment by mikebell90** [comment]
That would work but would require a lot of changes for developers. It’s unlikely to be anything other than a long tail. Regardless that was just one use case. Given we have probably several hundred databases the use case remains if for nothing more than it would be nice to log

**Comment by eulerto** [comment]
I'm closing this one because #1488 added a parameter that can be used to inform it is PgBouncer and not Postgres. There is also another issue  #653 that aims to advertise PgBouncer version via ParameterStatus message.
